src/fsharp/Optimizer.fs

```diff
@@ -2495,8 +2574,10 @@ and AddValEqualityInfo g m (v: ValRef) info =
     // ValValue is information that v = v2, where v2 does not change 
     // So we can't record this information for mutable values. An exception can be made
     // for "outArg" values arising from method calls since they are only temporarily mutable
-    // when their address is passed to the method call.
-    if v.IsMutable && not (v.IsCompilerGenerated && v.DisplayName.StartsWith(PrettyNaming.outArgCompilerGeneratedName)) then 
+    // when their address is passed to the method call. Another exception are mutable variables
+    // created for tuple elimination in branching tuple bindings because they are assigned to
+    // exactly once.
+    if v.IsMutable && not (v.IsCompilerGenerated && (v.DisplayName.StartsWith(PrettyNaming.outArgCompilerGeneratedName) || v.DisplayName.EndsWith(PrettyNaming.tempTupleElementAssignmentTargetName))) then 
```

Comment:
I'm really not happy with this, but just like in the outArg case, I need to indicate to this function that even though the Val is mutable, its ValueInfo should not be discarded. Is there a better way?

Reply:
We could set a flag on the Val.  Though to be honest I don't mind the technique you're using, and flags are precious (we only have 64-bit of flags for all future)

Reply:
Got it. I'll leave it as it is then.

Reply:
@kerams, would it make sense for `v.DisplayName.EndsWith(PrettyNaming.tempTupleElementAssignmentTargetName)` condition, to have it in a function that is kind of a dual of `MakeMutableStructuralBindingForTupleElement` rather than having the raw `.DisplayName.EndsWith` in that condition?

If the condition needs to be used in several places, looking up the `.EndsWith` and usages of the `PrettyNaming.tempTupleElementAssignmentTargetName` feels a bit wobbly.

This refactor, if applicable, would allow to hide the string from the PrettyNaming module, and just have the two functions in the optimizer instead, to keep it more self contained / implementation detail only.

Reply:
That sounds reasonable.

Reply:
@smoothdeveloper, see changes.

Reply:
@kerams, thanks, it looks great!